src/group_theory/exponent.lean

```diff
@@ -150,11 +200,69 @@ begin
     simp only [true_and, set.mem_univ, finset.coe_univ] at hg,
     exact ne_of_gt (order_of_pos g) hg },
   { intro g,
-    have h : (order_of g) ∣ (finset.univ : finset H).lcm order_of,
+    have h : (order_of g) ∣ (finset.univ : finset G).lcm order_of,
     { apply finset.dvd_lcm,
       exact finset.mem_univ g },
     cases h with m hm,
     rw [hm, pow_mul, pow_order_of_eq_one, one_pow] },
 end
 
+@[to_additive]
+lemma exponent_ne_zero_of_fintype [fintype G] : exponent G ≠ 0 :=
+by simpa [←lcm_order_eq_exponent, finset.lcm_eq_zero_iff] using λ x, (order_of_pos x).ne'
+
+end left_cancel_monoid
+
+section comm_monoid
+
+variable [cancel_comm_monoid G]
+
+open finset
+
+-- TODO: this is true without the `fintype`.
+@[to_additive]
+lemma exponent_eq_Sup_order_of [fintype G] : exponent G = Sup (set.range (order_of : G → ℕ)) :=
+begin
+  rw [←fintype.coe_image_univ, eq_comm,
+      nonempty.cSup_eq_max' (⟨1, by simp⟩ : ((@univ G _).image order_of).nonempty)],
+  generalize_proofs hne,
+  have := max'_mem (image order_of univ) hne,
+  rw mem_image at this,
+  obtain ⟨t, -, ht⟩ := this,
+  apply nat.dvd_antisymm,
+  { rw ←ht,
+    apply order_dvd_exponent },
+  refine nat.dvd_of_factors_subperm (exponent_ne_zero_of_fintype) _,
+  rw list.subperm_ext_iff,
+  by_contra' h,
+  obtain ⟨p, hp, hpe⟩ := h,
+  haveI hp := fact.mk (nat.prime_of_mem_factors hp),
+  simp only [←padic_val_nat_eq_factors_count p] at hpe,
+  set k := padic_val_nat p (order_of t) with hk,
+  obtain ⟨g, hg⟩ := exists_max_prime_pow_dvd_exponent G hp.1,
+  suffices : order_of t < order_of (t ^ (p ^ k) * g),
+  { rw ht at this,
+    exact this.not_le (le_max' _ _ $ by simp) },
+  have hpk  : p ^ k ∣ order_of t := pow_padic_val_nat_dvd,
+  have hpk' : order_of (t ^ p ^ k) = order_of t / p ^ k,
+  { rw [order_of_pow t, nat.gcd_eq_right hpk] },
+  obtain ⟨a, ha⟩ := nat.exists_eq_add_of_lt hpe,
+  have hcoprime : (order_of (t ^ p ^ k)).coprime (order_of g),
+  { rw [hg, nat.coprime_pow_right_iff, nat.coprime_comm],
+    apply or.resolve_right (nat.coprime_or_dvd_of_prime hp.1 _),
+    nth_rewrite 0 ←pow_one p,
+    convert pow_succ_padic_val_nat_not_dvd (order_of_pos (t ^ p ^ k)),
+    rw [hpk', padic_val_nat.div_pow hpk, hk, nat.sub_self],
+    { apply_instance },
+    rw ha,
+    exact nat.succ_pos _ },
+  rw [(commute.all _ g).order_of_mul_eq_mul_order_of_of_coprime hcoprime, hpk', hg, ha, ←ht, ←hk,
```

Comment:
Maybe it is worth to have a `order_of_mul_eq_mul_order_of_of_coprime'` as an alternative version for the order of multiplication reversed?

Reply:
I'm happy either way; I don't mind rw-ing `coprime_comm`. I think one of these `lcm order_of` lemmas is duplicated, fwiw!